File: app/src/main/java/com/originb/inkwisenote/data/entities/noterelationdata/NoteRelation.java
package com.originb.inkwisenote.data.entities.noterelationdata;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.PrimaryKey;
import com.originb.inkwisenote.data.entities.notedata.AtomicNoteEntity;
import com.originb.inkwisenote.data.entities.notedata.SmartBookEntity;
import lombok.Getter;
import lombok.Setter;

import java.util.Objects;

@Getter
@Setter
@Entity(tableName = "note_relation",
        foreignKeys = {
                @ForeignKey(entity = AtomicNoteEntity.class,
                        parentColumns = "note_id",
                        childColumns = "note_id",
                        onDelete = ForeignKey.CASCADE),
                @ForeignKey(entity = SmartBookEntity.class,
                        parentColumns = "book_id",
                        childColumns = "book_id",
                        onDelete = ForeignKey.CASCADE),
                @ForeignKey(entity = AtomicNoteEntity.class,
                        parentColumns = "note_id",
                        childColumns = "related_note_id",
                        onDelete = ForeignKey.CASCADE),
                @ForeignKey(entity = SmartBookEntity.class,
                        parentColumns = "book_id",
                        childColumns = "related_book_id",
                        onDelete = ForeignKey.CASCADE)
        })
public class NoteRelation {

    @PrimaryKey(autoGenerate = true)
    private int id;

    @ColumnInfo(name = "note_id")
    private Long noteId;

    @ColumnInfo(name = "book_id")
    private Long bookId;

    @ColumnInfo(name = "related_note_id")
    private Long relatedNoteId;

    @ColumnInfo(name = "related_book_id")
    private Long relatedBookId;

    @ColumnInfo(name = "relation_type")
    private Integer relationType;

    public NoteRelation(Long noteId, Long relatedNoteId, Long bookId, Long relatedBookId, Integer relationType) {
        this.noteId = noteId;
        this.relatedNoteId = relatedNoteId;
        this.bookId = bookId;
        this.relatedBookId = relatedBookId;

        this.relationType = relationType;
    }

    // Custom equals that compares only 'id'
    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true; // Reference equality
        if (obj == null || getClass() != obj.getClass()) return false; // Type check
        NoteRelation that = (NoteRelation) obj;
        return Objects.equals(noteId, that.noteId)
                && Objects.equals(relatedNoteId, that.relatedNoteId)
                && Objects.equals(bookId, that.bookId)
                && Objects.equals(relatedBookId, that.relatedBookId);
    }

    // Custom hashCode that considers only 'id'
    @Override
    public int hashCode() {
        return Objects.hash(noteId, relatedNoteId, bookId, relatedBookId); // Hash based on 'id'
    }
}

File: app/src/main/java/com/originb/inkwisenote2/modules/noterelation/data/NoteRelation.kt
package com.originb.inkwisenote2.modules.noterelation.data

import androidx.room.*
import com.originb.inkwisenote2.modules.smartnotes.data.AtomicNoteEntity
import com.originb.inkwisenote2.modules.smartnotes.data.SmartBookEntity
import java.util.*

@Entity(
    tableName = "note_relation",
    foreignKeys = [ForeignKey(
        entity = AtomicNoteEntity::class,
        parentColumns = "note_id",
        childColumns = "note_id",
        onDelete = ForeignKey.Companion.CASCADE
    ), ForeignKey(
        entity = SmartBookEntity::class,
        parentColumns = "book_id",
        childColumns = "book_id",
        onDelete = ForeignKey.Companion.CASCADE
    ), ForeignKey(
        entity = AtomicNoteEntity::class,
        parentColumns = "note_id",
        childColumns = "related_note_id",
        onDelete = ForeignKey.Companion.CASCADE
    ), ForeignKey(
        entity = SmartBookEntity::class,
        parentColumns = "book_id",
        childColumns = "related_book_id",
        onDelete = ForeignKey.Companion.CASCADE
    )],
    indices = [Index("note_id"), Index("book_id"), Index("related_note_id"), Index("related_book_id")]
)
class NoteRelation(
    @JvmField @field:ColumnInfo(name = "note_id") var noteId: Long,
    @JvmField @field:ColumnInfo(name = "related_note_id") var relatedNoteId: Long,
    @JvmField @field:ColumnInfo(
        name = "book_id"
    ) var bookId: Long,
    @JvmField @field:ColumnInfo(name = "related_book_id") var relatedBookId: Long,
    @JvmField @field:ColumnInfo(
        name = "relation_type"
    ) var relationType: Int
) {
    // Getters and Setters
    @JvmField
    @PrimaryKey(autoGenerate = true)
    var id: Int = 0

    // Custom equals that compares only 'id'
    override fun equals(obj: Any?): Boolean {
        if (this === obj) return true // Reference equality

        if (obj == null || javaClass != obj.javaClass) return false // Type check

        val that = obj as NoteRelation
        return noteId == that.noteId
                && relatedNoteId == that.relatedNoteId
                && bookId == that.bookId
                && relatedBookId == that.relatedBookId
    }

    // Custom hashCode that considers only 'id'
    override fun hashCode(): Int {
        return Objects.hash(noteId, relatedNoteId, bookId, relatedBookId) // Hash based on 'id'
    }
}
